code/espurna/rpnrules.cpp

```diff
@@ -124,52 +140,214 @@ void _rpnBrokerStatus(const String& topic, unsigned char id, unsigned int value)
+        return 'A';
+    case rpn_stack_value::Type::Value:
+    default:
+        return ' ';
+    }
+}
 
 #if RELAY_SUPPORT
 
-bool _rpnRelayStatus(rpn_context & ctxt, bool force) {
-    float status, id;
+rpn_error _rpnRelayStatus(rpn_context & ctxt, bool force) {
+    rpn_value id;
+    rpn_value status;
+
     rpn_stack_pop(ctxt, id);
     rpn_stack_pop(ctxt, status);
 
-    if (int(status) == 2) {
-        relayToggle(int(id));
-    } else if (force || (relayStatusTarget(int(id)) != (int(status) == 1))) {
-        relayStatus(int(id), int(status) == 1);
+    rpn_uint value = status.toUint();
+    if (value == 2) {
+        relayToggle(id.toUint());
+    } else if (relayStatusTarget(id.toUint()) != (value == 1)) {
+        relayStatus(id.toUint(), value == 1);
     }
-    return true;
+
+    return 0;
 }
 
+#endif // RELAY_SUPPORT
+
+#if RF_SUPPORT
+
+struct rpn_rfbridge_code {
+    String raw;
+    uint32_t hits;
+    decltype(millis()) last;
+    decltype(millis()) triggered;
+};
+
+static std::list<rpn_rfbridge_code> _rfb_codes;
+
+rpn_error _rpnRfbSequence(rpn_context& ctxt) {
+    rpn_value second;
+    rpn_stack_pop(ctxt, second);
+
+    rpn_value first;
+    rpn_stack_pop(ctxt, first);
+
+    String raw[2] {first.toString(), second.toString()};
+    rpn_rfbridge_code* refs[2] {nullptr, nullptr};
+
+    for (auto& recent : _rfb_codes) {
+        for (int index = 0; index < 2; ++index) {
+            refs[index] = (raw[index] == recent.raw) ? &recent : nullptr;
+        }
+    }
+
+    if ((refs[0] == nullptr) || (refs[1] == nullptr)) {
+        return rpn_operator_error::CannotContinue;
+    }
+
+    // purge codes to avoid matching again on the next rules run
+    // 'triggered' would be kind of quirky here, since we don't limit ourselves with any window
+    if ((millis() - refs[0]->last) > (millis() - refs[1]->last)) {
+        _rfb_codes.remove_if([&refs](rpn_rfbridge_code& code) {
+            return (refs[0] == &code) || (refs[1] == &code);
+        });
+        return rpn_operator_error::Ok;
+    }
+
+    return rpn_operator_error::CannotContinue;
+}
+
+rpn_error _rpnRfbMatcher(rpn_context& ctxt) {
+    rpn_value code;
+    rpn_stack_pop(ctxt, code);
+
+    rpn_value hits;
+    rpn_stack_pop(ctxt, hits);
+
+    String raw_code = code.toString();
+    auto result = std::find_if(_rfb_codes.begin(), _rfb_codes.end(), [&raw_code](const rpn_rfbridge_code& code) {
+        return code.raw == raw_code;
+    });
+
+    if (result == _rfb_codes.end()) {
+        return rpn_operator_error::CannotContinue;
+    }
+
+    // only process recent codes, ignore when rule is processing outside of this small window
+    if (millis() - (*result).last >= 2000) {
+        return rpn_operator_error::CannotContinue;
+    }
+
+    // hits == 1 is a single click, hits == 5 is long click
+    // we sort-of can distinguish single and double via timestamp, but it has a **very** unreliable timing
+    if ((*result).hits > hits.toUint()) {
```

Comment:
This check and the one below need to be slightly reworked to only trigger themselves when 'hits' stayed the same for a specific interval. After that, we could throw out the code instead of keeping it.

Another apparent issue with the check is eq vs. gt difference, we must have the exact number of hits or rule does not work. Perhaps the operator could push 'hits' on the stack when value stabilizes, and comparison could happen in the expression (including greater-than, greater-than-or-equal, etc.)

Reply:
*note-to-self:*

Yep, seems like a solution. So:
- broker from rf keeps sending codes as-is
- add additional task that checks if code had been stable for long enough, flag it as such
- rpn_match will only work with flagged codes, pushing number of hits
- after rules are done, code is removed from tracking (for the time being)

Other option was to keep checking it in the rule itself, simply making _rpn_run=true after match did not work. It is probably slower though